### ISSUE TITLE
cookies will require attribute value for SameSite attribute soon

### ISSUE THREAD

**Issue by uwedeportivo** [sourcegraph/sourcegraph#6167]
A customer running Sourcegraph under a different domain got this warning.

<img width="1973" alt="SameSiteSecure" src="https://user-images.githubusercontent.com/534011/67405516-b9dab480-f569-11e9-95bb-6fd5e9722262.png">

My guess the Sourcegraph instance references sourcegraph.com which sets cookies.

Modern browsers will soon require this. Chrome probably very soon. Our cookies coming from sourcegraph.com (docs, extensions ?) need to have SameSite="None" and "Secure".

**Comment by uwedeportivo** [comment]
Deadline for Chrome is Chrome 80 with release date 02/2020

**Comment by uwedeportivo** [comment]
https://www.chromium.org/updates/same-site

**Comment by uwedeportivo** [comment]
steps to reproduce:

- start local instance
- make sure you're signed out of sourcegraph.com (you can sign out now)
- install an extension

**Comment by uwedeportivo** [comment]
Chrome 80 beta doesn't enforce it (still accepts the cookies). Not sure if this will stand or will change before release of Chrome 80 in February

**Comment by uwedeportivo** [comment]
Safari and Safari Technology Preview (nightly builds) accept the cookies.

**Comment by uwedeportivo** [comment]
https://web.dev/samesite-cookies-explained/

**Comment by uwedeportivo** [comment]
![image](https://user-images.githubusercontent.com/534011/72293733-fe299e00-3608-11ea-873a-aecf523dcf61.png)

i enabled this upcoming behavior in chrome and tested and i could still install extensions and didn't see other adverse behavior.

this might not be as urgent as i originally thought.

we could whitelist certain routes in our router from requiring and setting a session. i'll discuss with the team

**Comment by felixfbecker** [comment]
Is there any _disadvantage_ to making the cookie SameSite? At least to get rid of the log spam

**Comment by uwedeportivo** [comment]
@felixfbecker SameSite is the attribute name. you would need to give me the value :-)

SameSite=None or SameSite=Lax or SameSite=Strict, etc...

for example if we make it Strict for all requests then we cannot use extensions coming from somewhere other than your current Sourcegraph instance (so you cannot install extensions from sourcegraph.com, so no good).

if we're making it None then we're opening ourselves up to cross site attacks.

different requests need different values of the SameSite attribute, we need to be careful how we implement this.

**Comment by uwedeportivo** [comment]
* https://www.troyhunt.com/promiscuous-cookies-and-their-impending-death-via-the-samesite-policy/
* https://web.dev/samesite-cookies-explained/

**Comment by felixfbecker** [comment]
Btw, if you have a proposal for how to do native integration auth better without cross-site cookies, that is a possibility too! E.g. using the native integration auth through OAuth to auth to the Sourcegraph instance. We had one case in the past where it didn't work because a user had cross-site cookies disabled. I haven't gotten around to think about this deeply yet because there are too many other things.

**Comment by uwedeportivo** [comment]
@felixfbecker good idea. i'll open a new issue for myself to go down that path and close this one once my PR lands

**Comment by uwedeportivo** [comment]
see https://github.com/sourcegraph/sourcegraph/issues/8350

**Comment by felixfbecker** [comment]
We are still getting warnings in the browser console:

```
🖥  Browser console: ⚠ warning A cookie associated with a resource at http://localhost/ was set with `SameSite=None` but without `Secure`. A future release of Chrome will only deliver cookies marked `SameSite=None` if they are also marked `Secure`. You can review cookies in developer tools under Application>Storage>Cookies and see more details at https://www.chromestatus.com/feature/5633521622188032.
--
  | at http://localhost:7080/github.com/sourcegraph/go-diff@3f415a150aec0685cb81b73cc201e762e075006d
```

Is this safe to ignore because it only happens on `localhost` served over plain HTTP?

**Comment by slimsag** [comment]
This is not something that should be solved in our cookie policy. We should not choose a cookie policy based on not using HTTPS and/or using `localhost`, both of which make Chrome and other browsers imply numerous weird security restrictions.

In other words, this is an issue with the e2e tests choosing to start the server on `localhost` and without HTTPS. You may choose to ignore it, if it is not causing problems, or otherwise someone will need to update the e2e tests to use HTTPS and/or `127.0.0.1` or a valid DNS name instead of `localhost`.

**Comment by felixfbecker** [comment]
Thanks for the response @slimsag! Note this also applies to our dev environments, the e2e tests just use `./dev/start.sh`. So as long as `./dev/start.sh` doesn't use HTTPS, the e2e tests won't either. If I read the warning correctly this will break (our dev environments _and_ e2e tests) in the future. According to the linked notes this actually seems to already have been rolled out to a subset of Chrome users that has been gradually increased: 

> **February, 2020**: Enforcement rollout for Chrome 80 Stable: The SameSite-by-default and ***SameSite=None-requires-Secure*** behaviors will begin rolling out to Chrome 80 Stable for an initial limited population starting the week of February 17, 2020, excluding the US President’s Day holiday on Monday. We will be closely monitoring and evaluating ecosystem impact from this initial limited phase through gradually increasing rollouts. 
>
> **March 2, 2020**: The enablement of the SameSite enforcements has been increased beyond the initial population. However, it is still targeting an overall limited global population of users on Chrome 80 stable and newer. We continue to monitor metrics and ecosystem feedback via our tracking bug, and other support channels.
>
> **March 9, 2020**: The rollout population has been additionally increased, although it continues to target a fraction of the overall Chrome 80 stable population. We continue to monitor metrics and ecosystem feedback via our tracking bug, and other support channels.

https://www.chromium.org/updates/same-site

**Comment by uwedeportivo** [comment]
it's annoying but safe to ignore.  it complains that the cookie wasn't marked "Secure". our middleware marks cookies secure iff we're on https (i know, duh).  i'm not 100% sure but i think chrome 80 and beyond will make an exception for localhost (but spam the console), so it won't break dev/start.sh.

i agree with stephen, i'm not comfortable changing the middleware according to dev/non-dev env vars. that seems dangerous.

we should probably switch to https for dev/start.sh. joe and i already did it with caddy. it's a little fiddly at the moment (so not quite turnkey) but it's doable.

**Comment by slimsag** [comment]
Looking closer, I recognize this is our session cookie and a more important aspect than both our e2e and dev environments exists here: how this will non-HTTPS Sourcegraph instances such as those created by someone following the docs on docs.sourcegraph.com.

> A future release of Chrome will only deliver cookies marked `SameSite=None` if they are also marked `Secure`.

It appears to me from this wording that our session cookie will not be delivered, even to the originating domain, which would mean sign in would be broken in all non-HTTPS Sourcegraph instances, as I understand it.

> i'm not 100% sure but i think chrome 80 and beyond will make an exception for localhost (but spam the console)

I couldn't find anything saying this is true, the only thing I could find is this auth0 docs entry indicating they will not make an exception for localhost https://auth0.com/docs/sessions/concepts/cookie-attributes#changes-you-need-to-make

**Comment by felixfbecker** [comment]
PR to ignore in e2e tests in https://github.com/sourcegraph/sourcegraph/pull/8917

Making our dev envs use HTTPS sounds like a good idea in any case. I think @eseliger also hacked something together for that in the past (using nginx? with a cool local address like `https://sourcegraph.test`).

**Comment by slimsag** [comment]
See https://web.dev/samesite-cookies-explained/#samesitenone-must-be-secure for how to test this new behavior in Chrome.

An initial assessment tells me we should do the following:

- When using HTTPS, use `SameSite=None` and `Secure` (as we do today)
  - Browser extension and sign in would work as expected.
- When using HTTP, use `SameSite=Lax` and not `Secure`.
  - This means the browser extension and other domains will NOT be able to make requests to Sourcegraph as they rely on cookie based authentication currently: https://github.com/sourcegraph/sourcegraph/issues/8350
  - But at least users/admins will be able to sign in, which would allow configuring HTTPS to get the browser extension to work.

The better long-term approach would be to see if we can get all Sourcegraph environments using HTTPS.

**Comment by eseliger** [comment]
I used https://github.com/FiloSottile/mkcert and forwarded all `*.test => 127.0.0.1` using dnsmasq

**Comment by felixfbecker** [comment]
> - When using HTTP, use SameSite=Lax and not Secure.
>   - This means the browser extension and other domains will NOT be able to make requests to Sourcegraph as they rely on cookie based authentication currently: #8350

Well if we do this, seems like we _have_ to make dev envs use HTTPS too. Otherwise, how would we the browser extension locally?

**Comment by slimsag** [comment]
> Well if we do this, seems like we have to make dev envs use HTTPS too. Otherwise, how would we the browser extension locally?

That is correct, you would not be able to use the browser extension locally without HTTPS.

**Comment by felixfbecker** [comment]
Just to be clear: In that case I consider making `./dev/start.sh` always use HTTPS a blocker to making that change.

**Comment by slimsag** [comment]
I want to be 100% clear here, we have two choices according to this message:

> A future release of Chrome will only deliver cookies marked SameSite=None if they are also marked Secure.

1. It becomes impossible to sign into all non-HTTPS Sourcegraph instances.
2. It remains possible to sign into all non-HTTPS Sourcegraph instances, but the browser extension cannot be used with them.

This is vastly more important that just our ability in the immediate term to test the browser extension easily locally. We're talking about production Sourcegraph instances that are ran without HTTPS behind firewalls no longer being accessible at all.

I agree we will want to make dev/start.sh use HTTPS -- but that is definitely not a blocker here given the implication that not making this change will mean users being locked out from their Sourcegraph instances.

**Comment by slimsag** [comment]
But also just to be clear again, I DO want the browser extension easily
testable and consider that high priority. It just may be that must come a
bit later.
-- 
Follow me on twitter @slimsag <https://twitter.com/slimsag>.

**Comment by felixfbecker** [comment]
I agree with that assessment. Luckily Chrome doesn't seem to have enabled this _yet_, so **while it is not enabled yet**, given that it is quite critical for our web team to be able to make changes to our own product as well (e.g. we are currently working on some things that are critical to make https://app.hubspot.com/contacts/2762526/company/407948923 successful) I would propose we make an effort to get both issues solved **together**. Both are important enough to warrant the effort imo, and should both be solvable together in the time frame before Chrome shuts us off. If Chrome happens to do this earlier than expected of course, I agree 100% that we should save customer instances asap, before ourselves.

**Comment by felixfbecker** [comment]
Do you agree with that? Or do you think we won't have enough time before Chrome flips the switch to do both?

**Comment by uwedeportivo** [comment]
damn, i'm beating myself up for missing this. of course this is bad for sourcegraph instances running on http and people won't be able to log in. @slimsag is right, this is getting high priority

**Comment by slimsag** [comment]
> damn, i'm beating myself up for missing this. 

Eh, the complex number of deployment configurations here is to blame, not yourself. I missed it too [when I commented above](https://github.com/sourcegraph/sourcegraph/issues/6167#issuecomment-597137129).

@felixfbecker I definitely want you guys to be able to continue developing. If we don't get this out in at least 3.14, customers will be at risk. From what I see in the timeline you quoted above this is already enforced for some large portion of Chrome users. Customers that lag behind are on versions as old as 3.10, and even up to date customers can lag behind up to a month. So to reduce the impact here this fix must go out in 3.14, and ideally in a patch before that.

What we can do is add a `BROWSER_EXTENSION_DEV_HACK=1` which is set in dev/start.sh that continues using the `SameSite=None` value and that'll give us as long as Chrome continues to support that behavior to add HTTPS to dev mode.

**Comment by uwedeportivo** [comment]
i will write up a handbook entry on how to run dev with https.

**Comment by uwedeportivo** [comment]
https://github.com/sourcegraph/about/pull/649

**Comment by slimsag** [comment]
Implementation plan:

- Use https://web.dev/samesite-cookies-explained/#samesitenone-must-be-secure to test your changes.
- When using HTTPS, make us use `SameSite=None` and `Secure` (as we do today)
  - Browser extension and sign in will continue to work as expected.
- When using HTTP, make us use `SameSite=Lax` (not what we do today) and not set `Secure` on the cookie (we do this today).
  - This means the browser extension and other domains will NOT be able to make requests to Sourcegraph as they rely on cookie based authentication currently: https://github.com/sourcegraph/sourcegraph/issues/8350 -- we should document that Sourcegraph instances without HTTPS may no longer be able to use the browser extension in Chrome due to this change of theirs and call that out in the changelog.
  - But at least users/admins will be able to sign in, which would allow configuring HTTPS to get the browser extension to work.
- The new behavior should be added to these two locations: https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+SameSite+lang:go&patternType=literal
- We do not need to do anything described above for working around the browser extension dev problem; Uwe added instructions for running dev with HTTPS so browser extension devs can use that.

**Comment by felixfbecker** [comment]
Agreeing with what was said above, and thanks @uwedeportivo for the fast HTTPS instructions!
I think this actually also the reason why the Puppeteer upgrade in https://github.com/sourcegraph/sourcegraph/pull/6966 was failing, because it uses a nightly version of Chrome that likely already prevents this. The good thing about that is that it means we will have e2e test coverage of this scenario once we upgrade!

**Comment by felixfbecker** [comment]
Also this lead me to believe that  https://github.com/sourcegraph/sourcegraph/issues/6167 is caused by this too (it all makes sense now!)
